File: MergeSort/js/MergeSort.js
'use strict';

class MergeSort {
  sort(listToSort) {
    if (1 < listToSort.length){
      var mid = this.findMiddle(listToSort),
        leftHalf = this.sort(listToSort.slice(0,mid)),
        rightHalf = this.sort(listToSort.slice(mid));
        listToSort = this.merge(leftHalf, rightHalf);
    }

    return listToSort;
  }

  merge(leftHalf, rightHalf) {
    var i=0,j=0,k=0,
      leftHalfLen = Array.isArray(leftHalf) ? leftHalf.length : 1,
      rightHalfLen = Array.isArray(rightHalf) ? rightHalf.length : 1,
      fullListLength = leftHalf.length + rightHalf.length,
      mergedList = [];

    while( k < fullListLength) {
      if (leftHalf[i] < rightHalf[j]) {
        mergedList.push(leftHalf[i]);
        i++;
      } else if(rightHalf[j] <= leftHalf[i]) {
        mergedList.push(rightHalf[j]);
        j++;
      }

      k++;
    }

    while (i < leftHalf.length) {
      mergedList.push(leftHalf[i]);
      i++;
      k++;
    }

    while (j < rightHalf.length) {
      mergedList.push(rightHalf[j]);
      j++;
      k++;
    }

    return mergedList;
  }

  findMiddle(listToSort) {
    return Math.ceil(listToSort.length/2);
  }
}

File: MergeSort/py/MergeSort.py
#!/usr/local/bin/python3

class MergeSort:
    def sort(self,alist):
        if len(alist) > 1:
            mid = len(alist)//2
            leftHalf = alist[:mid]
            rightHalf = alist[mid:]

            self.sort(leftHalf)
            self.sort(rightHalf)

            self.merge(alist, leftHalf,rightHalf)

        return alist

    def merge(self,alist,leftHalf,rightHalf):
        i = 0
        j = 0
        k = 0
        self.result = leftHalf + rightHalf

        while i < len(leftHalf) and j < len(rightHalf):
            if leftHalf[i] < rightHalf[j]:
                alist[k] = leftHalf[i]
                i = i + 1
            else:
                alist[k] = rightHalf[j]
                j = j + 1
            k = k + 1

        while i < len(leftHalf):
            alist[k] = leftHalf[i]
            i = i + 1
            k = k + 1

        while j < len(rightHalf):
            alist[k] = rightHalf[j]
            j = j + 1
            k = k + 1
